audio/drivers/coreaudio.cpp

```diff
@@ -0,0 +1,384 @@
+    if(status) {
+        qDebug("Failed to set render quality");
+        goto error;
+    }
+    desiredFormat.mSampleRate = getSampleRate(*audioDevice);
+    desiredFormat.mBytesPerPacket = sizeof(float) * 2;
+    desiredFormat.mBytesPerFrame = sizeof(float) * 2;
+    desiredFormat.mChannelsPerFrame = 2;
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_StreamFormat, kAudioUnitScope_Input, 0, &desiredFormat, sizeof(desiredFormat));
+    if(status) {
+        qDebug("Failed to set stream format");
+        goto error;
+    }
+    
+    AURenderCallbackStruct renderCallbackData;
+    renderCallbackData.inputProc = CoreAudioIOProc;
+    renderCallbackData.inputProcRefCon = ca;
+    
+    status = AudioUnitSetProperty(*audioUnit, kAudioUnitProperty_SetRenderCallback, kAudioUnitScope_Output, 0, &renderCallbackData, sizeof(renderCallbackData));
+    if(status) {
+        qDebug("Failed to set render callback");
+        goto error;
+    }
+    
+    AudioUnitInitialize(*audioUnit);
+    
+    return true;
+   error:
+    if(*audioUnit) {
+        AudioComponentInstanceDispose(*audioUnit);
+        *audioUnit = NULL;
+    }
+    *audioDevice = kAudioDeviceUnknown;
+    return false;
+}
+
+static OSStatus CoreAudioIOProc( void *inRefCon,
+                               AudioUnitRenderActionFlags *ioActionFlags,
+                               const AudioTimeStamp *inTimeStamp,
+                               UInt32 inBusNumber,
+                               UInt32 inNumberFrames,
+                               AudioBufferList *ioData)
+{
+      seq->setInitialMillisecondTimestampWithLatency();
+
+      long bytesPerFrame = sizeof(float) * ioData->mBuffers[0].mNumberChannels;
+
+      float* buffer = static_cast<float*>(ioData->mBuffers[0].mData);
+      UInt64 frames = ioData->mBuffers[0].mDataByteSize / bytesPerFrame;
+
+      seq->process((unsigned)frames, buffer);
+
+      return noErr;
+}
+
+bool CoreAudio::init(bool)
+      {
+      
+      CFRunLoopRef theRunLoop = NULL;
+      AudioObjectPropertyAddress theAddress = { kAudioHardwarePropertyRunLoop, kAudioObjectPropertyScopeGlobal, kAudioObjectPropertyElementMaster };
+      OSStatus osErr = AudioObjectSetPropertyData (kAudioObjectSystemObject, &theAddress, 0, NULL, sizeof(CFRunLoopRef), &theRunLoop);
+      if (osErr != noErr) {
+            initialized = false;
+            return false;
+      }
+  
+      AudioObjectPropertyAddress outputDeviceAddress = {
+            kAudioHardwarePropertyDefaultOutputDevice,
+            kAudioObjectPropertyScopeGlobal,
+            kAudioObjectPropertyElementMaster
+      };
+
+      AudioObjectAddPropertyListener(kAudioObjectSystemObject, &outputDeviceAddress, &onDefaultOutputDeviceChanged, this);
+
+      AudioDeviceID deviceId = getDefaultOutputDevice();
+      return openAudioUnit(&audioUnit, &deviceId, this);
```

Comment:
You still need to add
```
      _sampleRate = getSampleRate(deviceId);
```
before this return statement in order for notes to sound at the correct pitch.

Reply:
Ah yes. There's a couple places I need to change for this to work. See lines 167-171. Should `kAudioUnitScope_Input` be `kAudioUnitScope_Output`?

Reply:
I think desired output format should probably change with every output device change, yes?

Reply:
9c91143 adds the `_sampleRate = getSampleRate(deviceId);`

I was not able to duplicate your issue in any consistent and useful way on my machine. What are you running this build on?

Reply:
MacBook Pro (15 inch, 2018), macOS Catalina Version 10.15.4.

Reply:
With your recent changes you can remove the line in question. It is no longer necessary because of what you have done elsewhere.

Reply:
Ok. Is your pitch issue fixed, or are you still having issues?

Reply:
I am no longer having issues. The issue is fixed. And it is still fixed even after I remove the line that I asked you to add.